src/main/java/net/mcreator/generator/GeneratorTemplatesList.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * MCreator (https://mcreator.net/)
+ * Copyright (C) 2012-2020, Pylo
+ * Copyright (C) 2020-2021, Pylo, opensource contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package net.mcreator.generator;
+
+import net.mcreator.element.GeneratableElement;
+
+import java.io.File;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A generator templates list is used for generating several similar templates for each item on a list provided by given
+ * mod element.
+ *
+ * @param groupName The name of this group of templates shown in workspace panel.
+ * @param listData  The collection used by workspace to generate given templates.
+ * @param element   The {@link GeneratableElement} used to process tokens in template names.
+ * @param templates The map of templates to be generated for each entry of {@code listData};
+ *                  keys are templates themselves, values represent generation conditions of their key template
+ *                  for all items on the mentioned collection.
+ */
+public record GeneratorTemplatesList(String groupName, List<?> listData, GeneratableElement element,
+									 Map<GeneratorTemplate, List<Boolean>> templates) {
+
+	/**
+	 * Replaces tokens on name of provided list template with appropriate values from
+	 * {@link GeneratorTemplatesList#element} and returns the target file.
+	 *
+	 * @param generatorTemplate One of templates in this list instance.
+	 * @param index             Index of listData element for which to acquire the output file path.
+	 * @return The file generated from given list template with given token values.
+	 */
+	public File processTokens(GeneratorTemplate generatorTemplate, int index) {

```

Comment:
As GeneratorTemplate already exists in `templates` Map, it may be cleaner to just take the index as parameter for the method and get key by index from `templates`. This would require to change type from Map to LinkedHashMap so the order is properly preserved.

Because right now, one could _technically_ provide unmatching GeneratorTemplate and index.

EDIT: LinkedHashMap may be a bit unclean too. Maybe splitting Map into two lists: GeneratorTemplate and conditions list. Given list data and templates size must match, requesting the same from three lists may not be that wrong.

Reply:
Method may also be called `getFileForTemplateIndex` as it describes better what it does, as it does not only process tokens, but also creates a file

Reply:
>Given list data and templates size must match

This is not how I intended it to be. The current implementation of this class would look like this:
|Group name|Template №1|Template №2|Template №3|
|-|-|-|-|
|List item №1|:x:|:heavy_check_mark:|:heavy_check_mark:|
|List item №2|:x:|:x:|:heavy_check_mark:|
|List item №3|:heavy_check_mark:|:x:|:heavy_check_mark:|
|List item №4|:x:|:heavy_check_mark:|:x:|

Reply:
I don't think I understand the table, but I thought each listData entry corresponds to one GeneratorTemplate for a given GeneratorTemplatesList

Reply:
Ok nevermind, correct me if I got it wrong.

Each template in the list of templates in GeneratorTemplatesList can use every (or not, it is optional from user perspective) listData index?

If this is the case, then maybe just rename the method and just add a check that generatorTemplate provided in the argument exists in templates Map and if not, throw RuntimeException as processing unrelated template would violate the logic

Reply:
Seeing that in both use places of `processTokens` you generate list of all files, maybe restructuring this method to `getAllFiles()` that returns list of files for given GeneratorTemplatesList would be a solution that would eliminate the need for checks

Reply:
>Each template in the list of templates in GeneratorTemplatesList can use every (or not, it is optional from user perspective) listData index?

It uses every index for which condition defined for the template object passes.

Reply:
>...maybe restructuring this method to `getAllFiles()` that returns list of files for given GeneratorTemplatesList would be a solution...

These 2 iterations are ran in different ways (list data first and templates second in ME code menu, the other way in generator).

Reply:
Is there a particular reason for different orders? Technically getAllFiles should return list of files that works for either generator or user preview

Reply:
Code menu adds separator between files generated for each item in `listData`.

Reply:
Then keep it as is, just rename the method to `getFileForTemplateAndIndex(GeneratorTemplate, int)`, maybe you could also move condition checking to `isConditionForTemplateAndIndex(GeneratorTemplate, int)` as seeing `if (fileList.templates().get(modElementListFile).get(i)) {` outside GeneratorTemplatesList can be a bit confusing

Reply:
`fileList.templates().keySet()` could also have a method in GeneratorTemplatesList, maybe `getTemplates`

Reply:
I'll implement this tomorrow.

Reply:
Ok, no worries, take your time :)

Reply:
✔️ Fixed